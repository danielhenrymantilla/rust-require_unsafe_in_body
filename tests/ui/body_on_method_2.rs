#[macro_use]
extern crate require_unsafe_in_body;

#[derive(Clone, Copy, Default)]
struct Foo<T> (
    ::core::marker::PhantomData<T>,
);

trait MyRead {
    unsafe
    fn my_read<U> (self: &'_ Self, p: *const U) -> U
    where
        U : Copy,
    ;
}

impl<T> MyRead for Foo<T> {
    #[require_unsafe_in_body]
    unsafe
    fn my_read<U> (self: &'_ Self, p: *const U) -> U
    where
        U : Copy,
    {
        unsafe {
            ::core::ptr::read(p)
        }
    }
}

fn main ()
{
    let x = 42;
    let foo: Foo<()> = Foo::default();
    assert_eq!(
        unsafe { foo.my_read(&x) },
        42,
    );
}
