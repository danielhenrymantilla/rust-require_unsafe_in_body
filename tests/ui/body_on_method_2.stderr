error: `#[require_unsafe_in_body]` does not support directly decorating a method; you need to decorate the whole `impl` or `trait` block with `#[require_unsafe_in_bodies]`
  --> $DIR/body_on_method_2.rs:18:5
   |
18 |     #[require_unsafe_in_body]
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0046]: not all trait items implemented, missing: `my_read`
  --> $DIR/body_on_method_2.rs:17:1
   |
10 | /     unsafe
11 | |     fn my_read<U> (self: &'_ Self, p: *const U) -> U
12 | |     where
13 | |         U : Copy,
14 | |     ;
   | |_____- `my_read` from trait
...
17 |   impl<T> MyRead for Foo<T> {
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^ missing `my_read` in implementation

For more information about this error, try `rustc --explain E0046`.
